[TEST-FAILURE] MapReduceTest.testNullFromObjectCombiner

https://hazelcast-l337.ci.cloudbees.com/job/Hazelcast-3.x-OpenJDK8/com.hazelcast$hazelcast/258/testReport/com.hazelcast.mapreduce/MapReduceTest/testNullFromObjectCombiner/

```
java.lang.AssertionError: expected:<1225> but was:<1110>
    at org.junit.Assert.fail(Assert.java:88)
    at org.junit.Assert.failNotEquals(Assert.java:743)
    at org.junit.Assert.assertEquals(Assert.java:118)
    at org.junit.Assert.assertEquals(Assert.java:144)
    at com.hazelcast.mapreduce.MapReduceTest.testNullFromObjectCombiner(MapReduceTest.java:726)
```

--- a/hazelcast-client/src/test/java/com/hazelcast/client/mapreduce/AbstractClientMapReduceJobTest.java
+++ b/hazelcast-client/src/test/java/com/hazelcast/client/mapreduce/AbstractClientMapReduceJobTest.java
@@ -20,8 +20,9 @@ import com.hazelcast.client.config.ClientConfig;
 import com.hazelcast.client.config.XmlClientConfigBuilder;
 import com.hazelcast.config.Config;
 import com.hazelcast.config.XmlConfigBuilder;
+import com.hazelcast.test.HazelcastTestSupport;
 
-public abstract class AbstractClientMapReduceJobTest {
+public abstract class AbstractClientMapReduceJobTest extends HazelcastTestSupport {
 
     protected ClientConfig buildClientConfig() {
         ClientConfig config = new XmlClientConfigBuilder().build();

--- a/hazelcast-client/src/test/java/com/hazelcast/client/mapreduce/ClientMapReduceTest.java
+++ b/hazelcast-client/src/test/java/com/hazelcast/client/mapreduce/ClientMapReduceTest.java
@@ -62,13 +62,16 @@ public class ClientMapReduceTest
     }
 
     @Test(timeout = 60000, expected = ExecutionException.class)
-    public void testExceptionDistribution()
-            throws Exception {
+    public void testExceptionDistribution() throws Exception {
         Config config = buildConfig();
 
         HazelcastInstance h1 = Hazelcast.newHazelcastInstance(config);
         HazelcastInstance h2 = Hazelcast.newHazelcastInstance(config);
         HazelcastInstance h3 = Hazelcast.newHazelcastInstance(config);
+
+        assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         HazelcastInstance client = HazelcastClient.newHazelcastClient(null);
         IMap<Integer, Integer> m1 = client.getMap(MAP_NAME);
@@ -92,13 +95,16 @@ public class ClientMapReduceTest
     }
 
     @Test(timeout = 60000, expected = CancellationException.class)
-    public void testInProcessCancellation()
-            throws Exception {
+    public void testInProcessCancellation() throws Exception {
         Config config = buildConfig();
 
         HazelcastInstance h1 = Hazelcast.newHazelcastInstance(config);
         HazelcastInstance h2 = Hazelcast.newHazelcastInstance(config);
         HazelcastInstance h3 = Hazelcast.newHazelcastInstance(config);
+
+        assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         HazelcastInstance client = HazelcastClient.newHazelcastClient(null);
         IMap<Integer, Integer> m1 = client.getMap(MAP_NAME);
@@ -123,13 +129,16 @@ public class ClientMapReduceTest
     }
 
     @Test(timeout = 60000)
-    public void testMapper()
-            throws Exception {
+    public void testMapper() throws Exception {
         Config config = buildConfig();
 
         HazelcastInstance h1 = Hazelcast.newHazelcastInstance(config);
         HazelcastInstance h2 = Hazelcast.newHazelcastInstance(config);
         HazelcastInstance h3 = Hazelcast.newHazelcastInstance(config);
+
+        assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         HazelcastInstance client = HazelcastClient.newHazelcastClient(null);
         IMap<Integer, Integer> m1 = client.getMap(MAP_NAME);
@@ -150,13 +159,16 @@ public class ClientMapReduceTest
     }
 
     @Test(timeout = 60000)
-    public void testMapperReducer()
-            throws Exception {
+    public void testMapperReducer() throws Exception {
         Config config = buildConfig();
 
         HazelcastInstance h1 = Hazelcast.newHazelcastInstance(config);
         HazelcastInstance h2 = Hazelcast.newHazelcastInstance(config);
         HazelcastInstance h3 = Hazelcast.newHazelcastInstance(config);
+
+        assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         HazelcastInstance client = HazelcastClient.newHazelcastClient(null);
         IMap<Integer, Integer> m1 = client.getMap(MAP_NAME);
@@ -184,13 +196,16 @@ public class ClientMapReduceTest
     }
 
     @Test(timeout = 60000)
-    public void testMapperCollator()
-            throws Exception {
+    public void testMapperCollator() throws Exception {
         Config config = buildConfig();
 
         HazelcastInstance h1 = Hazelcast.newHazelcastInstance(config);
         HazelcastInstance h2 = Hazelcast.newHazelcastInstance(config);
         HazelcastInstance h3 = Hazelcast.newHazelcastInstance(config);
+
+        assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         HazelcastInstance client = HazelcastClient.newHazelcastClient(null);
         IMap<Integer, Integer> m1 = client.getMap(MAP_NAME);
@@ -212,13 +227,16 @@ public class ClientMapReduceTest
     }
 
     @Test(timeout = 60000)
-    public void testKeyedMapperCollator()
-            throws Exception {
+    public void testKeyedMapperCollator() throws Exception {
         Config config = buildConfig();
 
         HazelcastInstance h1 = Hazelcast.newHazelcastInstance(config);
         HazelcastInstance h2 = Hazelcast.newHazelcastInstance(config);
         HazelcastInstance h3 = Hazelcast.newHazelcastInstance(config);
+
+        assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         HazelcastInstance client = HazelcastClient.newHazelcastClient(null);
         IMap<Integer, Integer> m1 = client.getMap(MAP_NAME);
@@ -236,13 +254,16 @@ public class ClientMapReduceTest
     }
 
     @Test(timeout = 60000)
-    public void testKeyPredicateMapperCollator()
-            throws Exception {
+    public void testKeyPredicateMapperCollator() throws Exception {
         Config config = buildConfig();
 
         HazelcastInstance h1 = Hazelcast.newHazelcastInstance(config);
         HazelcastInstance h2 = Hazelcast.newHazelcastInstance(config);
         HazelcastInstance h3 = Hazelcast.newHazelcastInstance(config);
+
+        assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         HazelcastInstance client = HazelcastClient.newHazelcastClient(null);
         IMap<Integer, Integer> m1 = client.getMap(MAP_NAME);
@@ -261,13 +282,16 @@ public class ClientMapReduceTest
     }
 
     @Test(timeout = 60000)
-    public void testMapperReducerCollator()
-            throws Exception {
+    public void testMapperReducerCollator() throws Exception {
         Config config = buildConfig();
 
         HazelcastInstance h1 = Hazelcast.newHazelcastInstance(config);
         HazelcastInstance h2 = Hazelcast.newHazelcastInstance(config);
         HazelcastInstance h3 = Hazelcast.newHazelcastInstance(config);
+
+        assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         HazelcastInstance client = HazelcastClient.newHazelcastClient(null);
         IMap<Integer, Integer> m1 = client.getMap(MAP_NAME);
@@ -294,13 +318,16 @@ public class ClientMapReduceTest
     }
 
     @Test(timeout = 60000)
-    public void testAsyncMapper()
-            throws Exception {
+    public void testAsyncMapper() throws Exception {
         Config config = buildConfig();
 
         HazelcastInstance h1 = Hazelcast.newHazelcastInstance(config);
         HazelcastInstance h2 = Hazelcast.newHazelcastInstance(config);
         HazelcastInstance h3 = Hazelcast.newHazelcastInstance(config);
+
+        assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         HazelcastInstance client = HazelcastClient.newHazelcastClient(null);
         IMap<Integer, Integer> m1 = client.getMap(MAP_NAME);
@@ -338,13 +365,16 @@ public class ClientMapReduceTest
     }
 
     @Test(timeout = 60000)
-    public void testKeyedAsyncMapper()
-            throws Exception {
+    public void testKeyedAsyncMapper() throws Exception {
         Config config = buildConfig();
 
         HazelcastInstance h1 = Hazelcast.newHazelcastInstance(config);
         HazelcastInstance h2 = Hazelcast.newHazelcastInstance(config);
         HazelcastInstance h3 = Hazelcast.newHazelcastInstance(config);
+
+        assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         HazelcastInstance client = HazelcastClient.newHazelcastClient(null);
         IMap<Integer, Integer> m1 = client.getMap(MAP_NAME);
@@ -382,13 +412,16 @@ public class ClientMapReduceTest
     }
 
     @Test(timeout = 60000)
-    public void testAsyncMapperReducer()
-            throws Exception {
+    public void testAsyncMapperReducer() throws Exception {
         Config config = buildConfig();
 
         HazelcastInstance h1 = Hazelcast.newHazelcastInstance(config);
         HazelcastInstance h2 = Hazelcast.newHazelcastInstance(config);
         HazelcastInstance h3 = Hazelcast.newHazelcastInstance(config);
+
+        assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         HazelcastInstance client = HazelcastClient.newHazelcastClient(null);
         IMap<Integer, Integer> m1 = client.getMap(MAP_NAME);
@@ -433,13 +466,16 @@ public class ClientMapReduceTest
     }
 
     @Test(timeout = 60000)
-    public void testAsyncMapperCollator()
-            throws Exception {
+    public void testAsyncMapperCollator() throws Exception {
         Config config = buildConfig();
 
         HazelcastInstance h1 = Hazelcast.newHazelcastInstance(config);
         HazelcastInstance h2 = Hazelcast.newHazelcastInstance(config);
         HazelcastInstance h3 = Hazelcast.newHazelcastInstance(config);
+
+        assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         HazelcastInstance client = HazelcastClient.newHazelcastClient(null);
         IMap<Integer, Integer> m1 = client.getMap(MAP_NAME);
@@ -483,13 +519,16 @@ public class ClientMapReduceTest
     }
 
     @Test(timeout = 60000)
-    public void testAsyncMapperReducerCollator()
-            throws Exception {
+    public void testAsyncMapperReducerCollator() throws Exception {
         Config config = buildConfig();
 
         HazelcastInstance h1 = Hazelcast.newHazelcastInstance(config);
         HazelcastInstance h2 = Hazelcast.newHazelcastInstance(config);
         HazelcastInstance h3 = Hazelcast.newHazelcastInstance(config);
+
+        assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         HazelcastInstance client = HazelcastClient.newHazelcastClient(null);
         IMap<Integer, Integer> m1 = client.getMap(MAP_NAME);

--- a/hazelcast-client/src/test/java/com/hazelcast/client/mapreduce/DistributedMapperClientMapReduceTest.java
+++ b/hazelcast-client/src/test/java/com/hazelcast/client/mapreduce/DistributedMapperClientMapReduceTest.java
@@ -59,13 +59,16 @@ public class DistributedMapperClientMapReduceTest
     }
 
     @Test(timeout = 30000)
-    public void testMapperReducer()
-            throws Exception {
+    public void testMapperReducer() throws Exception {
         Config config = buildConfig();
 
         HazelcastInstance h1 = Hazelcast.newHazelcastInstance(config);
         HazelcastInstance h2 = Hazelcast.newHazelcastInstance(config);
         HazelcastInstance h3 = Hazelcast.newHazelcastInstance(config);
+
+        assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         HazelcastInstance client = HazelcastClient.newHazelcastClient(null);
         IMap<Integer, Integer> m1 = client.getMap(MAP_NAME);
@@ -93,13 +96,16 @@ public class DistributedMapperClientMapReduceTest
     }
 
     @Test(timeout = 30000)
-    public void testMapperReducerCollator()
-            throws Exception {
+    public void testMapperReducerCollator() throws Exception {
         Config config = buildConfig();
 
         HazelcastInstance h1 = Hazelcast.newHazelcastInstance(config);
         HazelcastInstance h2 = Hazelcast.newHazelcastInstance(config);
         HazelcastInstance h3 = Hazelcast.newHazelcastInstance(config);
+
+        assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         HazelcastInstance client = HazelcastClient.newHazelcastClient(null);
         IMap<Integer, Integer> m1 = client.getMap(MAP_NAME);
@@ -126,13 +132,16 @@ public class DistributedMapperClientMapReduceTest
     }
 
     @Test(timeout = 30000)
-    public void testAsyncMapperReducer()
-            throws Exception {
+    public void testAsyncMapperReducer() throws Exception {
         Config config = buildConfig();
 
         HazelcastInstance h1 = Hazelcast.newHazelcastInstance(config);
         HazelcastInstance h2 = Hazelcast.newHazelcastInstance(config);
         HazelcastInstance h3 = Hazelcast.newHazelcastInstance(config);
+
+        assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         HazelcastInstance client = HazelcastClient.newHazelcastClient(null);
         IMap<Integer, Integer> m1 = client.getMap(MAP_NAME);
@@ -180,13 +189,16 @@ public class DistributedMapperClientMapReduceTest
     }
 
     @Test(timeout = 30000)
-    public void testAsyncMapperReducerCollator()
-            throws Exception {
+    public void testAsyncMapperReducerCollator() throws Exception {
         Config config = buildConfig();
 
         HazelcastInstance h1 = Hazelcast.newHazelcastInstance(config);
         HazelcastInstance h2 = Hazelcast.newHazelcastInstance(config);
         HazelcastInstance h3 = Hazelcast.newHazelcastInstance(config);
+
+        assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         HazelcastInstance client = HazelcastClient.newHazelcastClient(null);
         IMap<Integer, Integer> m1 = client.getMap(MAP_NAME);

--- a/hazelcast-client/src/test/java/com/hazelcast/client/mapreduce/aggregation/AbstractAggregationTest.java
+++ b/hazelcast-client/src/test/java/com/hazelcast/client/mapreduce/aggregation/AbstractAggregationTest.java
@@ -41,8 +41,11 @@ public class AbstractAggregationTest
 
     @BeforeClass
     public static void startup() {
-        Hazelcast.newHazelcastInstance();
-        Hazelcast.newHazelcastInstance();
+        HazelcastInstance h1 = Hazelcast.newHazelcastInstance();
+        HazelcastInstance h2 = Hazelcast.newHazelcastInstance();
+
+        assertClusterSizeEventually(2, h1);
+        assertClusterSizeEventually(2, h2);
 
         HAZELCAST_INSTANCE = HazelcastClient.newHazelcastClient();
     }

--- a/hazelcast/src/test/java/com/hazelcast/mapreduce/CustomDataSourceMapReduceTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/mapreduce/CustomDataSourceMapReduceTest.java
@@ -65,6 +65,8 @@ public class CustomDataSourceMapReduceTest
         final HazelcastInstance h3 = nodeFactory.newHazelcastInstance();
 
         assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         JobTracker jobTracker = h1.getJobTracker("default");
         Job<String, Integer> job = jobTracker.newJob(new CustomKeyValueSource());

--- a/hazelcast/src/test/java/com/hazelcast/mapreduce/DistributedMapperMapReduceTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/mapreduce/DistributedMapperMapReduceTest.java
@@ -41,12 +41,15 @@ public class DistributedMapperMapReduceTest
     private static final String MAP_NAME = "default";
 
     @Test(timeout = 30000)
-    public void testMapperReducer()
-            throws Exception {
+    public void testMapperReducer() throws Exception {
         TestHazelcastInstanceFactory nodeFactory = createHazelcastInstanceFactory(4);
         HazelcastInstance h1 = nodeFactory.newHazelcastInstance();
         HazelcastInstance h2 = nodeFactory.newHazelcastInstance();
         HazelcastInstance h3 = nodeFactory.newHazelcastInstance();
+
+        assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         IMap<Integer, Integer> m1 = h1.getMap(MAP_NAME);
         for (int i = 0; i < 100; i++) {
@@ -73,12 +76,15 @@ public class DistributedMapperMapReduceTest
     }
 
     @Test(timeout = 30000)
-    public void testMapperCustomIntermediateCombinerReducer()
-            throws Exception {
+    public void testMapperCustomIntermediateCombinerReducer() throws Exception {
         TestHazelcastInstanceFactory nodeFactory = createHazelcastInstanceFactory(4);
         HazelcastInstance h1 = nodeFactory.newHazelcastInstance();
         HazelcastInstance h2 = nodeFactory.newHazelcastInstance();
         HazelcastInstance h3 = nodeFactory.newHazelcastInstance();
+
+        assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         IMap<Integer, Integer> m1 = h1.getMap(MAP_NAME);
         for (int i = 0; i < 100; i++) {
@@ -106,12 +112,15 @@ public class DistributedMapperMapReduceTest
     }
 
     @Test(timeout = 30000)
-    public void testMapperReducerCollator()
-            throws Exception {
+    public void testMapperReducerCollator() throws Exception {
         TestHazelcastInstanceFactory nodeFactory = createHazelcastInstanceFactory(4);
         HazelcastInstance h1 = nodeFactory.newHazelcastInstance();
         HazelcastInstance h2 = nodeFactory.newHazelcastInstance();
         HazelcastInstance h3 = nodeFactory.newHazelcastInstance();
+
+        assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         IMap<Integer, Integer> m1 = h1.getMap(MAP_NAME);
         for (int i = 0; i < 100; i++) {
@@ -137,12 +146,15 @@ public class DistributedMapperMapReduceTest
     }
 
     @Test(timeout = 30000)
-    public void testAsyncMapperReducer()
-            throws Exception {
+    public void testAsyncMapperReducer() throws Exception {
         TestHazelcastInstanceFactory nodeFactory = createHazelcastInstanceFactory(4);
         HazelcastInstance h1 = nodeFactory.newHazelcastInstance();
         HazelcastInstance h2 = nodeFactory.newHazelcastInstance();
         HazelcastInstance h3 = nodeFactory.newHazelcastInstance();
+
+        assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         IMap<Integer, Integer> m1 = h1.getMap(MAP_NAME);
         for (int i = 0; i < 100; i++) {
@@ -186,12 +198,15 @@ public class DistributedMapperMapReduceTest
     }
 
     @Test(timeout = 30000)
-    public void testAsyncMapperReducerCollator()
-            throws Exception {
+    public void testAsyncMapperReducerCollator() throws Exception {
         TestHazelcastInstanceFactory nodeFactory = createHazelcastInstanceFactory(4);
         HazelcastInstance h1 = nodeFactory.newHazelcastInstance();
         HazelcastInstance h2 = nodeFactory.newHazelcastInstance();
         HazelcastInstance h3 = nodeFactory.newHazelcastInstance();
+
+        assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         IMap<Integer, Integer> m1 = h1.getMap(MAP_NAME);
         for (int i = 0; i < 100; i++) {

--- a/hazelcast/src/test/java/com/hazelcast/mapreduce/ListSetMapReduceTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/mapreduce/ListSetMapReduceTest.java
@@ -41,9 +41,7 @@ public class ListSetMapReduceTest
         extends HazelcastTestSupport {
 
     @Test(timeout = 60000)
-    public void testMapReduceWithList()
-            throws Exception {
-
+    public void testMapReduceWithList() throws Exception {
         TestHazelcastInstanceFactory nodeFactory = createHazelcastInstanceFactory(3);
 
         final HazelcastInstance h1 = nodeFactory.newHazelcastInstance();
@@ -51,6 +49,8 @@ public class ListSetMapReduceTest
         final HazelcastInstance h3 = nodeFactory.newHazelcastInstance();
 
         assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         int expectedResult = 0;
         IList<Integer> list = h1.getList("default");
@@ -78,21 +78,16 @@ public class ListSetMapReduceTest
 
     @Test(timeout = 60000)
     @Category(ProblematicTest.class)
-    public void testMapReduceWithSet()
-            throws Exception {
-
+    public void testMapReduceWithSet() throws Exception {
         TestHazelcastInstanceFactory nodeFactory = createHazelcastInstanceFactory(3);
 
         final HazelcastInstance h1 = nodeFactory.newHazelcastInstance();
         final HazelcastInstance h2 = nodeFactory.newHazelcastInstance();
         final HazelcastInstance h3 = nodeFactory.newHazelcastInstance();
 
-        assertTrueEventually(new AssertTask() {
-            @Override
-            public void run() {
-                assertEquals(3, h1.getCluster().getMembers().size());
-            }
-        });
+        assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         int expectedResult = 0;
         ISet<Integer> set = h1.getSet("default");

--- a/hazelcast/src/test/java/com/hazelcast/mapreduce/MapReduceManagedContextTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/mapreduce/MapReduceManagedContextTest.java
@@ -47,9 +47,7 @@ public class MapReduceManagedContextTest
     private static final String MAP_NAME = "default";
 
     @Test
-    public void testManagedContextMapper()
-            throws Exception {
-
+    public void testManagedContextMapper() throws Exception {
         TestHazelcastInstanceFactory nodeFactory = createHazelcastInstanceFactory(3);
 
         final HazelcastInstance h1 = nodeFactory.newHazelcastInstance();
@@ -57,6 +55,8 @@ public class MapReduceManagedContextTest
         final HazelcastInstance h3 = nodeFactory.newHazelcastInstance();
 
         assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         IMap<String, Integer> m1 = h1.getMap(MAP_NAME);
         for (int i = 0; i < 100; i++) {

--- a/hazelcast/src/test/java/com/hazelcast/mapreduce/MapReduceTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/mapreduce/MapReduceTest.java
@@ -57,9 +57,7 @@ public class MapReduceTest
     private static final String MAP_NAME = "default";
 
     @Test(timeout = 30000)
-    public void testPartitionPostpone()
-            throws Exception {
-
+    public void testPartitionPostpone() throws Exception {
         TestHazelcastInstanceFactory nodeFactory = createHazelcastInstanceFactory(3);
 
         final HazelcastInstance h1 = nodeFactory.newHazelcastInstance();
@@ -67,6 +65,8 @@ public class MapReduceTest
         final HazelcastInstance h3 = nodeFactory.newHazelcastInstance();
 
         assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         IMap<Integer, Integer> m1 = h1.getMap(MAP_NAME);
         for (int i = 0; i < 100; i++) {
@@ -88,9 +88,7 @@ public class MapReduceTest
     }
 
     @Test(timeout = 30000, expected = ExecutionException.class)
-    public void testExceptionDistributionWithCollator()
-            throws Exception {
-
+    public void testExceptionDistributionWithCollator() throws Exception {
         TestHazelcastInstanceFactory nodeFactory = createHazelcastInstanceFactory(3);
 
         final HazelcastInstance h1 = nodeFactory.newHazelcastInstance();
@@ -98,6 +96,8 @@ public class MapReduceTest
         final HazelcastInstance h3 = nodeFactory.newHazelcastInstance();
 
         assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         IMap<Integer, Integer> m1 = h1.getMap(MAP_NAME);
         for (int i = 0; i < 100; i++) {
@@ -127,9 +127,7 @@ public class MapReduceTest
     }
 
     @Test(timeout = 30000, expected = ExecutionException.class)
-    public void testExceptionDistribution()
-            throws Exception {
-
+    public void testExceptionDistribution() throws Exception {
         TestHazelcastInstanceFactory nodeFactory = createHazelcastInstanceFactory(3);
 
         final HazelcastInstance h1 = nodeFactory.newHazelcastInstance();
@@ -137,6 +135,8 @@ public class MapReduceTest
         final HazelcastInstance h3 = nodeFactory.newHazelcastInstance();
 
         assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         IMap<Integer, Integer> m1 = h1.getMap(MAP_NAME);
         for (int i = 0; i < 100; i++) {
@@ -159,9 +159,7 @@ public class MapReduceTest
     }
 
     @Test(timeout = 30000, expected = CancellationException.class)
-    public void testInProcessCancellation()
-            throws Exception {
-
+    public void testInProcessCancellation() throws Exception {
         TestHazelcastInstanceFactory nodeFactory = createHazelcastInstanceFactory(3);
 
         final HazelcastInstance h1 = nodeFactory.newHazelcastInstance();
@@ -169,6 +167,8 @@ public class MapReduceTest
         final HazelcastInstance h3 = nodeFactory.newHazelcastInstance();
 
         assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         IMap<Integer, Integer> m1 = h1.getMap(MAP_NAME);
         for (int i = 0; i < 100; i++) {
@@ -192,9 +192,7 @@ public class MapReduceTest
     }
 
     @Test(timeout = 30000)
-    public void testMapper()
-            throws Exception {
-
+    public void testMapper() throws Exception {
         TestHazelcastInstanceFactory nodeFactory = createHazelcastInstanceFactory(3);
 
         final HazelcastInstance h1 = nodeFactory.newHazelcastInstance();
@@ -202,6 +200,8 @@ public class MapReduceTest
         final HazelcastInstance h3 = nodeFactory.newHazelcastInstance();
 
         assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         IMap<Integer, Integer> m1 = h1.getMap(MAP_NAME);
         for (int i = 0; i < 100; i++) {
@@ -221,14 +221,16 @@ public class MapReduceTest
     }
 
     @Test(timeout = 30000)
-    public void testKeyedMapperCollator()
-            throws Exception {
-
+    public void testKeyedMapperCollator() throws Exception {
         TestHazelcastInstanceFactory nodeFactory = createHazelcastInstanceFactory(3);
 
         HazelcastInstance h1 = nodeFactory.newHazelcastInstance();
         HazelcastInstance h2 = nodeFactory.newHazelcastInstance();
         HazelcastInstance h3 = nodeFactory.newHazelcastInstance();
+
+        assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         IMap<Integer, Integer> m1 = h1.getMap(MAP_NAME);
         for (int i = 0; i < 10000; i++) {
@@ -245,14 +247,16 @@ public class MapReduceTest
     }
 
     @Test(timeout = 30000)
-    public void testKeyPredicateMapperCollator()
-            throws Exception {
-
+    public void testKeyPredicateMapperCollator() throws Exception {
         TestHazelcastInstanceFactory nodeFactory = createHazelcastInstanceFactory(3);
 
         HazelcastInstance h1 = nodeFactory.newHazelcastInstance();
         HazelcastInstance h2 = nodeFactory.newHazelcastInstance();
         HazelcastInstance h3 = nodeFactory.newHazelcastInstance();
+
+        assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         IMap<Integer, Integer> m1 = h1.getMap(MAP_NAME);
         for (int i = 0; i < 10000; i++) {
@@ -270,14 +274,16 @@ public class MapReduceTest
     }
 
     @Test(timeout = 30000)
-    public void testMapperComplexMapping()
-            throws Exception {
-
+    public void testMapperComplexMapping() throws Exception {
         TestHazelcastInstanceFactory nodeFactory = createHazelcastInstanceFactory(3);
 
         HazelcastInstance h1 = nodeFactory.newHazelcastInstance();
         HazelcastInstance h2 = nodeFactory.newHazelcastInstance();
         HazelcastInstance h3 = nodeFactory.newHazelcastInstance();
+
+        assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         IMap<Integer, Integer> m1 = h1.getMap(MAP_NAME);
         for (int i = 0; i < 100; i++) {
@@ -295,14 +301,16 @@ public class MapReduceTest
     }
 
     @Test(timeout = 30000)
-    public void testMapperReducer()
-            throws Exception {
-
+    public void testMapperReducer() throws Exception {
         TestHazelcastInstanceFactory nodeFactory = createHazelcastInstanceFactory(3);
 
         HazelcastInstance h1 = nodeFactory.newHazelcastInstance();
         HazelcastInstance h2 = nodeFactory.newHazelcastInstance();
         HazelcastInstance h3 = nodeFactory.newHazelcastInstance();
+
+        assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         IMap<Integer, Integer> m1 = h1.getMap(MAP_NAME);
         for (int i = 0; i < 100; i++) {
@@ -329,9 +337,7 @@ public class MapReduceTest
     }
 
     @Test(timeout = 60000)
-    public void testMapperReducerChunked()
-            throws Exception {
-
+    public void testMapperReducerChunked() throws Exception {
         TestHazelcastInstanceFactory nodeFactory = createHazelcastInstanceFactory(3);
 
         final HazelcastInstance h1 = nodeFactory.newHazelcastInstance();
@@ -339,6 +345,8 @@ public class MapReduceTest
         final HazelcastInstance h3 = nodeFactory.newHazelcastInstance();
 
         assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         final IMap<Integer, Integer> m1 = h1.getMap(MAP_NAME);
         for (int i = 0; i < 10000; i++) {
@@ -377,14 +385,16 @@ public class MapReduceTest
     }
 
     @Test(timeout = 30000)
-    public void testMapperCollator()
-            throws Exception {
-
+    public void testMapperCollator() throws Exception {
         TestHazelcastInstanceFactory nodeFactory = createHazelcastInstanceFactory(3);
 
         HazelcastInstance h1 = nodeFactory.newHazelcastInstance();
         HazelcastInstance h2 = nodeFactory.newHazelcastInstance();
         HazelcastInstance h3 = nodeFactory.newHazelcastInstance();
+
+        assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         IMap<Integer, Integer> m1 = h1.getMap(MAP_NAME);
         for (int i = 0; i < 100; i++) {
@@ -409,14 +419,16 @@ public class MapReduceTest
     }
 
     @Test(timeout = 30000)
-    public void testMapperReducerCollator()
-            throws Exception {
-
+    public void testMapperReducerCollator() throws Exception {
         TestHazelcastInstanceFactory nodeFactory = createHazelcastInstanceFactory(3);
 
         HazelcastInstance h1 = nodeFactory.newHazelcastInstance();
         HazelcastInstance h2 = nodeFactory.newHazelcastInstance();
         HazelcastInstance h3 = nodeFactory.newHazelcastInstance();
+
+        assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         IMap<Integer, Integer> m1 = h1.getMap(MAP_NAME);
         for (int i = 0; i < 100; i++) {
@@ -442,14 +454,16 @@ public class MapReduceTest
     }
 
     @Test(timeout = 30000)
-    public void testAsyncMapper()
-            throws Exception {
-
+    public void testAsyncMapper() throws Exception {
         TestHazelcastInstanceFactory nodeFactory = createHazelcastInstanceFactory(3);
 
         HazelcastInstance h1 = nodeFactory.newHazelcastInstance();
         HazelcastInstance h2 = nodeFactory.newHazelcastInstance();
         HazelcastInstance h3 = nodeFactory.newHazelcastInstance();
+
+        assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         IMap<Integer, Integer> m1 = h1.getMap(MAP_NAME);
         for (int i = 0; i < 100; i++) {
@@ -489,14 +503,16 @@ public class MapReduceTest
     }
 
     @Test(timeout = 30000)
-    public void testKeyedAsyncMapper()
-            throws Exception {
-
+    public void testKeyedAsyncMapper() throws Exception {
         TestHazelcastInstanceFactory nodeFactory = createHazelcastInstanceFactory(3);
 
         HazelcastInstance h1 = nodeFactory.newHazelcastInstance();
         HazelcastInstance h2 = nodeFactory.newHazelcastInstance();
         HazelcastInstance h3 = nodeFactory.newHazelcastInstance();
+
+        assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         IMap<Integer, Integer> m1 = h1.getMap(MAP_NAME);
         for (int i = 0; i < 100; i++) {
@@ -536,14 +552,16 @@ public class MapReduceTest
     }
 
     @Test(timeout = 30000)
-    public void testAsyncMapperReducer()
-            throws Exception {
-
+    public void testAsyncMapperReducer() throws Exception {
         TestHazelcastInstanceFactory nodeFactory = createHazelcastInstanceFactory(3);
 
         HazelcastInstance h1 = nodeFactory.newHazelcastInstance();
         HazelcastInstance h2 = nodeFactory.newHazelcastInstance();
         HazelcastInstance h3 = nodeFactory.newHazelcastInstance();
+
+        assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         IMap<Integer, Integer> m1 = h1.getMap(MAP_NAME);
         for (int i = 0; i < 100; i++) {
@@ -590,14 +608,16 @@ public class MapReduceTest
     }
 
     @Test(timeout = 30000)
-    public void testAsyncMapperCollator()
-            throws Exception {
-
+    public void testAsyncMapperCollator() throws Exception {
         TestHazelcastInstanceFactory nodeFactory = createHazelcastInstanceFactory(3);
 
         HazelcastInstance h1 = nodeFactory.newHazelcastInstance();
         HazelcastInstance h2 = nodeFactory.newHazelcastInstance();
         HazelcastInstance h3 = nodeFactory.newHazelcastInstance();
+
+        assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         IMap<Integer, Integer> m1 = h1.getMap(MAP_NAME);
         for (int i = 0; i < 100; i++) {
@@ -642,14 +662,16 @@ public class MapReduceTest
     }
 
     @Test(timeout = 30000)
-    public void testAsyncMapperReducerCollator()
-            throws Exception {
-
+    public void testAsyncMapperReducerCollator() throws Exception {
         TestHazelcastInstanceFactory nodeFactory = createHazelcastInstanceFactory(3);
 
         HazelcastInstance h1 = nodeFactory.newHazelcastInstance();
         HazelcastInstance h2 = nodeFactory.newHazelcastInstance();
         HazelcastInstance h3 = nodeFactory.newHazelcastInstance();
+
+        assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         IMap<Integer, Integer> m1 = h1.getMap(MAP_NAME);
         for (int i = 0; i < 100; i++) {
@@ -695,14 +717,16 @@ public class MapReduceTest
     }
 
     @Test(timeout = 30000)
-    public void testNullFromObjectCombiner()
-            throws Exception {
-
+    public void testNullFromObjectCombiner() throws Exception {
         TestHazelcastInstanceFactory nodeFactory = createHazelcastInstanceFactory(3);
 
         HazelcastInstance h1 = nodeFactory.newHazelcastInstance();
         HazelcastInstance h2 = nodeFactory.newHazelcastInstance();
         HazelcastInstance h3 = nodeFactory.newHazelcastInstance();
+
+        assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         IMap<Integer, Integer> m1 = h1.getMap(MAP_NAME);
         for (int i = 0; i < 100; i++) {
@@ -728,14 +752,16 @@ public class MapReduceTest
     }
 
     @Test(timeout = 30000)
-    public void testDataSerializableIntermediateObject()
-            throws Exception {
-
+    public void testDataSerializableIntermediateObject() throws Exception {
         TestHazelcastInstanceFactory nodeFactory = createHazelcastInstanceFactory(3);
 
         HazelcastInstance h1 = nodeFactory.newHazelcastInstance();
         HazelcastInstance h2 = nodeFactory.newHazelcastInstance();
         HazelcastInstance h3 = nodeFactory.newHazelcastInstance();
+
+        assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         IMap<Integer, Integer> m1 = h1.getMap(MAP_NAME);
         for (int i = 0; i < 100; i++) {

--- a/hazelcast/src/test/java/com/hazelcast/mapreduce/MultiMapMapReduceTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/mapreduce/MultiMapMapReduceTest.java
@@ -51,9 +51,7 @@ public class MultiMapMapReduceTest
     };
 
     @Test(timeout = 60000)
-    public void testMapReduceWithMultiMap()
-            throws Exception {
-
+    public void testMapReduceWithMultiMap() throws Exception {
         TestHazelcastInstanceFactory nodeFactory = createHazelcastInstanceFactory(3);
 
         final HazelcastInstance h1 = nodeFactory.newHazelcastInstance();
@@ -61,6 +59,8 @@ public class MultiMapMapReduceTest
         final HazelcastInstance h3 = nodeFactory.newHazelcastInstance();
 
         assertClusterSizeEventually(3, h1);
+        assertClusterSizeEventually(3, h2);
+        assertClusterSizeEventually(3, h3);
 
         MultiMap<Integer, Integer> multiMap = h1.getMultiMap("default");
         for (int i = 0; i < 1000; i++) {

--- a/hazelcast/src/test/java/com/hazelcast/mapreduce/aggregation/AbstractAggregationTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/mapreduce/aggregation/AbstractAggregationTest.java
@@ -42,7 +42,10 @@ public class AbstractAggregationTest
     public static void startup() {
         INSTANCE_FACTORY = new TestHazelcastInstanceFactory(2);
         HAZELCAST_INSTANCE = INSTANCE_FACTORY.newHazelcastInstance();
-        INSTANCE_FACTORY.newHazelcastInstance();
+        HazelcastInstance hazelcastInstance = INSTANCE_FACTORY.newHazelcastInstance();
+
+        assertClusterSizeEventually(2, HAZELCAST_INSTANCE);
+        assertClusterSizeEventually(2, hazelcastInstance);
     }
 
     @AfterClass

